File: cpp/tournament.cpp
/*
************************************************************
* Name:  Mariya Popova                                     *
* Project:  1 Pente C++                                    *
* Class: CMPS 366-01 Organization of Programming Languages *
* Date:  September 27, 2023                                *
************************************************************
*/

#include "tournament.h"

/* ********************************************************************* 
Function Name: coinToss
Purpose: simulate a coin toss, used at the start of a new round when 
            scores are tied
Parameters: none
Return Value: returns index of first player to play
Algorithm: 
        1) Generate a random value to simulate coin toss
        2) Announce winner
Assistance Received: none 
********************************************************************* */
int Tournament::coinToss(){
    string input;
    int index = 0;

    cout << "Coin toss to determine first move. Choose 'heads' or 'tails': ";
    cin >> input;

    // Random value generated, converted immediately to heads or tails
    string winner = (rand()%2 == 0 ? "heads" : "tails");      
    cout << winner << " won!" << endl;

    // Allowing user to input any capitaliztion
    for (auto& x : input) {
        x = tolower(x);
    }
        
    // If guessed incorrectly, computer goes first
    m_board.setCompNext(winner != input);
    if (winner != input){
        // Index of computer player
        cout << "Computer goes first!" << endl;
        index = 1;
    }
    else {
        cout << "Human goes first!" << endl;
    }

    return index;
}

/* ********************************************************************* 
Function Name: run
Purpose:  run the tournament, which is comprised of rounds; update
            tournament score at the end of each round
Parameters: 
        a_index - index either read from file or set to zero and to be determined 
                    if board is empty
Return Value: none
Algorithm:
        1) Get current scores from board and store in tournament scores
        2) Play rounds while human player says yes or until exit call
        3) If user didn't exit before end of round, scores displayed
        4) Tournament scores displayed and winner announed if user 
        said no to playing more rounds and there was no premature exit
Assistance Received: none 
********************************************************************* */
void Tournament::run(int a_index){
    Round round(m_board, m_roster);
    // Get current score from m_board and set it as tournament score
    // m_board score variables set to zero as they will track round score
    m_humanScore = m_board.getHumanScore();
    m_compScore = m_board.getCompScore();
    m_board.setCompScore(0);
    m_board.setHumanScore(0);

    string input;

    // Play rounds while human says yes or until exit call
    do {
        // If board if empty, first player is determined
        if (m_board.countNonEmpty() == 0){
            if (m_compScore == m_humanScore){
                a_index = coinToss();
            }
            else if (m_compScore > m_humanScore) {
                cout << "Computer plays first because computer has highest score." << endl;
                a_index = 1;
            }
            else {
                cout << "Human plays first because human has highest score." << endl;
                a_index = 0;
            }

            // Roster, next player and color updated
            m_roster[a_index]->setColor('W');
            m_roster[a_index^1]->setColor('B');
            m_board.setCompNext(a_index == 1);
            m_board.setBlackNext(false);
        }

        // Round starts
        round.play(a_index);     

        // If user didn't call to exit, round ends in a win/drawn 
        // and scores are updated and displayed
        if (!round.exitCall()){
            m_humanScore += m_board.getHumanScore();  
            m_compScore += m_board.getCompScore();
            cout << "Round score:" << endl;
            cout << "Human: " << m_board.getHumanScore() << endl;
            cout << "Computer: " << m_board.getCompScore() << endl;
            cout << "Winner of this round: ";
            if (m_board.getHumanScore() > m_board.getCompScore()){
                cout << "human player!" << endl;
            }
            else if (m_board.getHumanScore() <<m_board.getCompScore()){
                cout << "computer player!" << endl;
            }

            cout << endl << "Tournament scores:" << endl;
            cout << "Human: " << m_humanScore << endl;
            cout << "Computer: " << m_compScore << endl << endl; 

            m_board.boardInit(); 

            cout << "Play another round? Enter 'Y' or 'N': " << endl; 
            cin >> input;
   
            while (input != "Y" && input != "N" && input != "y" && input != "n"){
                cout << "Please enter 'Y' or 'N': ";
                cin >> input;
            }
            cout << endl;
            if (input == "n" || input == "N"){
                break;
            }
        }
    } while (!round.exitCall());
    
    // If user didn't call to exit prematurely, tournament scores and winner displayed
    if (!round.exitCall()){
        endTournament();
    }

    // Updated to be saved accurately into file
    m_board.setHumanScore(m_humanScore);
    m_board.setCompScore(m_compScore);
}


/* ********************************************************************* 
Function Name: endTournament
Purpose:   display tournament scores at the end of the tournament 
            and announce winner
Parameters: none
Return Value: none
Algorithm: none
Assistance Received: none 
********************************************************************* */
void Tournament::endTournament(){
    cout << "Tournament scores:" << endl;
    cout << "Human: " << m_humanScore << endl;
    cout << "Computer: " << m_compScore << endl << endl; 
    if (m_compScore > m_humanScore){
        cout << "The winner of the tournament is the computer player!" << endl;
    }
    else if (m_compScore < m_humanScore) {
        cout << "The winner of the tournament is the human player!" << endl;
    }
    else {
        cout << "The tournament ends in a draw!" << endl;
    }
}

File: prolog/tournament.pl
% *************************************************************
% * Name:  Mariya Popova                                      *
% * Project:  4 Pente Prolog                                  *
% * Class:  CMPS 366-01 Organization of Programming Languages *
% * Date:  December 5, 2023                                   *
% *************************************************************

% ********************************************************************* 
% Function Name: tournament
% Purpose: carry out the tournament, which is comprised of rounds; update
%          tournament score at the end of each round
% Parameters: 
%      board - board state passed from file or initialized 
%      human - list of human captures and tournament score 
%      comp - list of computer captures and tournament score 
%      nextplayer - list of next player and their stone color 
%      filename - file to save to if the user decides to end the game 
% Algorithm:
%      1. Store tournament scores before round changes value to track round score.
%      2. Store the current state of the round (values returned by round) to quit or continue 
%         the game accordingly.
%      3. If the user didnt quit before the end of the round, round scores displayed. 
%         If the user quit, save the current state to a file.
%      4. Tournament scores displayed, and the winner announced if the user 
%         says no to playing another round. Stats saved to a file.
%      5. The tournament is called again with updated tournament score values if
%         the user agrees to play another round from the prompt.
% Assistance Received: none 
% ********************************************************************* 
tournament(Board, [HCap, HScore], [CCap, CScore], NextPlayer, Filename):-
    % Store scores before round changes value to keep track of round score
    HumanTourScore = HScore,
    CompTourScore = CScore,
    round(Board, [HCap, HScore], [CCap, CScore], NextPlayer, [FinalBoard, [HRoundCap, HRoundScore], [CRoundCap, CRoundScore], RoundNextPlayer, ExitCase]),
    % Act according to exit case
    board_init(19, FreshBoard),
    (   ExitCase = 'quit' 
    ->  save_game(FinalBoard, [HRoundCap, HumanTourScore], [CRoundCap, CompTourScore], RoundNextPlayer, Filename),
        format("Game saved to file.~n")
    ;
        scores_after_round(FinalBoard, RoundNextPlayer, [HRoundCap, HRoundScore], [CRoundCap, CRoundScore], HumanTourScore, CompTourScore, TournamentScores),
        TournamentScores = [HumanScore, CompScore],
        format("Play another round? Enter 'Y' or 'N': ~n"),
        read_line_to_string(user_input, UserInput),
        string_upper(UserInput, Input),
        another_round_prompt(Input, FreshBoard, [0, HumanScore], [0, CompScore], NextPlayer, Filename)
    ).

% ********************************************************************* 
% Function Name: scores_after_round
% Purpose: compute round and tournament scores and display them
% Parameters: 
%      board - board after the round ends
%      nextplayer - list of the next player and their stone color 
%      human - list of human captures and round score 
%      comp - list of computer captures and round score 
%      humantourscore - humans tournament score before this round 
%      comptourscore - computers tournament score before this round 
%      tournamentscores - out variable, list of human and computer score after this round
% Algorithm:
%      1. Compute round scores by adding capture count, any 5 in a row scores, and counting
%         4 in a rows for each player.
%      2. Compute tournament scores by adding to the current tournament scores the round scores.
%      3. Print both, return a list of tournament scores to the tournament function.
% Assistance Received: none 
% ********************************************************************* 
scores_after_round(Board, NextPlayer, [HumanC, HumanS], [CompC, CompS], HumanTourScore, CompTourScore, TournamentScores) :-
    % Check the next player color to add 4 in a row counts to the correct players
    (NextPlayer = ["human", "white"] ; NextPlayer = ["computer", "black"]),
    count_4_in_a_row(Board, "b", 0, 0, Comp4),
    CompRoundScore is CompC + CompS + Comp4,
    count_4_in_a_row(Board, "w", 0, 0, Human4),
    HumanRoundScore is HumanC + HumanS + Human4,
    CompTournament is CompTourScore + CompRoundScore,
    HumanTournament is HumanTourScore + HumanRoundScore,
    format("Round scores:~n"),
    print_scores(HumanRoundScore, CompRoundScore),
    format("Tournament scores:~n"),
    print_scores(HumanTournament, CompTournament),
    TournamentScores = [HumanTournament, CompTournament].

scores_after_round(Board, NextPlayer, [HumanC, HumanS], [CompC, CompS], HumanTourScore, CompTourScore, TournamentScores) :-
    % Check the next player color to add 4 in a row counts to the correct players
    (NextPlayer = ["computer", "white"] ; NextPlayer = ["human", "black"]),
    count_4_in_a_row(Board, "w", 0, 0, Comp4),
    CompRoundScore is CompC + CompS + Comp4,
    count_4_in_a_row(Board, "b", 0, 0, Human4),
    HumanRoundScore is HumanC + HumanS + Human4,
    CompTournament is CompTourScore + CompRoundScore,
    HumanTournament is HumanTourScore + HumanRoundScore,
    format("Round scores:~n"),
    print_scores(HumanRoundScore, CompRoundScore),
    format("Tournament scores:~n"),
    print_scores(HumanTournament, CompTournament),
    TournamentScores = [HumanTournament, CompTournament].

% ********************************************************************* 
% Function Name: print_scores
% Purpose: score printing function
% Parameters: 
%      humanscore - score passed for the human 
%      compscore - score passed for the computer 
% Algorithm: none
% Assistance Received: none 
% ********************************************************************* 
print_scores(HumanScore, CompScore) :-
    format("Human: ~w~n", [HumanScore]),
    format("Computer: ~w~n", [CompScore]).

% ********************************************************************* 
% Function Name: another_round_prompt
% Purpose: get valid input from the user 
% Parameters: 
%      userinput - to decide which rule to follow
%      board - new board to save to file or start next rounf
%      human - list of human captures and tournament score 
%      comp - list of computer captures and tournament score 
%      nextplayer - list of the next player and their stone color 
%      filename - filename to save game to if no more rounds
% Algorithm: 
%      1. "y" - call tournament to start next round
%      2. "n" - display winner of tournament and save game to file
%      4. invalid input - prompt and call another_round_prompt again
% Assistance Received: none 
% ********************************************************************* 
another_round_prompt("N", FreshBoard, [0, HumanScore], [0, CompScore], NextPlayer, Filename):-
    writeln("Tournament scores:"),
    print_scores(HumanScore, CompScore),
    (   HumanScore > CompScore
    ->  format("Winner of the tournament: Human player!~n")
    ; 
        CompScore > HumanScore 
    ->  format("Winner of the tournament: Computer player!~n")
    ; 
        format("Winner of the tournament: It's a draw!~n")
    ),
    save_game(FreshBoard, [0, HumanScore], [0, CompScore], NextPlayer, Filename).
another_round_prompt("Y", FreshBoard, [0, HumanScore], [0, CompScore], _, Filename):-
    tournament(FreshBoard, [0, HumanScore], [0, CompScore], ["", ""], Filename).
another_round_prompt(_, FreshBoard, [0, HumanScore], [0, CompScore], NextPlayer, Filename):-
    format("Enter 'Y' or 'N': ~n"),
    read_line_to_string(user_input, UserInput),
    string_upper(UserInput, Input),
    another_round_prompt(Input, FreshBoard, [0, HumanScore], [0, CompScore], NextPlayer, Filename).
